### PR TITLE
INSP: add inspection that checks non-existent crates

### PR DESCRIPTION
**Pull request by Kobzol**
This PR add more basic infrastructure to TOML, based on the new local crate index. The first commit introduces a simple inspection which checks crates in `Cargo.toml` that do not exist (=were not found in the index). It uses the new TOML visitor (https://github.com/intellij-rust/intellij-rust/pull/6578) and also the dependency visitor from https://github.com/intellij-rust/intellij-rust/pull/3787. After another inspection is added, this dependency visitor should be extracted into a separate file.

The second commit then introduces tests for the inspection, which however require a lot of new infrastructure. Mainly it adds a service interface and a test implementation for the local crate index (taken from https://github.com/intellij-rust/intellij-rust/pull/5988).

Closes: https://github.com/intellij-rust/intellij-rust/pull/3787
Closes: https://github.com/intellij-rust/intellij-rust/pull/5988
Related: https://github.com/intellij-rust/intellij-rust/issues/6463

changelog: Add inspection that checks non-existent crates in Cargo.toml. Enabled only with the experimental local crate index.

### CODE REVIEW

**Comment by Kobzol** [comment]
We may need to add some method to the crate resolver interface for checking if the crate resolver is ready. Otherwise this inspection could result in false positives if `getCrate` returns null.

Btw, when running CLion locally with this PR, the local crate index sometimes does not have any entry e.g. for `serde` or `log`, and `getAllCrateNames` returns only 64 crate names, IDK why.

File: toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexService.kt [hunk 1]
```diff
@@ -337,3 +312,20 @@ private object CrateExternalizer : DataExternalizer<CargoRegistryCrate> {
         return CargoRegistryCrate(versions)
     }
 }
+
+private data class ParsedVersion(
```
Comment (avrong) [inline toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexService.kt:316]: Can we move it to CargoRegistryCrateVersion's companion as `ParsedVersion` has no use outside of the `crateFromJson` function?

**Review by avrong — CHANGES_REQUESTED** [review]

File: toml/src/test/kotlin/org/rust/toml/inspections/CrateNotFoundInspectionTest.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.rust.ide.experiments.RsExperiments
+import org.rust.ide.inspections.RsInspectionsTestBase
+import org.rust.openapiext.runWithEnabledFeature
+import org.rust.toml.crates.local.CargoRegistryCrate
+import org.rust.toml.crates.local.CargoRegistryCrateVersion
+import org.rust.toml.crates.local.withMockedCrates
+
+class CrateNotFoundInspectionTest : RsInspectionsTestBase(CrateNotFoundInspection::class) {
+    fun `test missing crate in dependencies`() = doTest(
+        """
```
[the remark this answers is not in the document]
Comment (Kobzol) [reply toml/src/test/kotlin/org/rust/toml/inspections/CrateNotFoundInspectionTest.kt:19]: Uh, no idea why it was formatted like that. Thanks.

File: toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexService.kt (hunk 1 above)
Comment (Kobzol) [reply toml/src/main/kotlin/org/rust/toml/crates/local/CratesLocalIndexService.kt:316]: I suppose that it should be in this file, as it's specific to the format of the local index? `CargoRegistryCrateVersion` should be a general interface describing a crate version. For example, if we later port the crates.io resolver to this service interface, it should use `CargoRegistryCrateVersion`, but it shouldn't have to know about `ParsedVersion`.

File: toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+                // [dependencies], [x86.dev-dependencies], etc.
+                if (dependencyNameIndex == names.lastIndex) {
+                    element.entries.forEach {
+                        val name = it.key.name ?: return@forEach
+                        val value = it.value ?: return@forEach
+                        if (value is TomlLiteral && value.kind is TomlLiteralKind.String) {
+                            handleDependency(Dependency(name, it.key, mapOf("version" to value)), holder)
+                        }
+                        else if (value is TomlInlineTable) {
```
Comment (avrong) [inline toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:41]:
```suggestion
                        } else if (value is TomlInlineTable) {
```

File: toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                            handleDependency(Dependency(name, it.key, collectProperties(value)), holder)
+                        }
+                    }
+                }
+                else if (dependencyNameIndex != -1) {
```
Comment (avrong) [inline toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:46]:
```suggestion
                } else if (dependencyNameIndex != -1) {
```

File: toml/src/main/kotlin/org/rust/toml/crates/local/TestCrateResolverService.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml.crates.local
+
+import org.jetbrains.annotations.TestOnly
+
+@TestOnly
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/crates/local/TestCrateResolverService.kt:10]:
```suggestion
```
This class is registered in XML so it should not be excluded. Otherwise, the compilation of other plugins depending on IJ Rust would be broken. See `TestCargoProjectsServiceImpl` for example

File: toml/src/main/resources/inspectionDescriptions/CrateNotFound.html [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<html>
+<body>
+Detects unknown dependency crates in Cargo.toml.
```
Comment (ortem) [inline toml/src/main/resources/inspectionDescriptions/CrateNotFound.html:3]:
```suggestion
Detects unknown dependency crates in Cargo.toml
```

File: toml/src/test/kotlin/org/rust/toml/inspections/CrateNotFoundInspectionTest.kt [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    """)
+
+    fun `test missing crate in specific dependency`() = doTest("""
+        [dependencies.<warning descr="Crate foo not found">foo</warning>]
+    """)
+
+    fun `test existing crate in dependencies`() = doTest("""
+        [dependencies]
+        foo = "1"
+    """, crate("foo", "1"))
```
Comment (ortem) [inline toml/src/test/kotlin/org/rust/toml/inspections/CrateNotFoundInspectionTest.kt:40]: Let's add more tests with several crates provided, both positive (with `<warning>`) and negative

File: toml/src/main/kotlin/org/rust/toml/crates/local/CrateResolverService.kt [hunk 8]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml.crates.local
+
+import com.intellij.openapi.components.service
+
+interface CrateResolverService {
```
[the remark this answers is not in the document]
Comment (Kobzol) [reply toml/src/main/kotlin/org/rust/toml/crates/local/CrateResolverService.kt:10]: Originally I thought that we also might implement this interface with the crates.io resolver. Is that futile?

[the remark this answers is not in the document]
Comment (ortem) [reply toml/src/main/kotlin/org/rust/toml/crates/local/CrateResolverService.kt:10]: Frankly, I'm not sure if this would be very useful in general since using crates.io API is slow and isn't suitable enough for real-time inspections in IDE. But if you know cases when such implementation would be nice to have, please feel free to describe them

[the remark this answers is not in the document]
Comment (Kobzol) [reply toml/src/main/kotlin/org/rust/toml/crates/local/CrateResolverService.kt:10]: I only wanted to do that to make the inspections work if the local (git) is not available for some reason. But I agree that its functionality (and latency) leaves a lot to be desired. I'll rename the service :)

File: toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    private fun handleDependency(dependency: Dependency, holder: ProblemsHolder) {
+        if (CratesLocalIndexService.getInstance().getCrate(dependency.name) == null) {
+            holder.registerProblem(dependency.nameElement, "Crate ${dependency.name} not found")
+        }
+    }
+}
+
+data class Dependency(val name: String, val nameElement: TomlKeySegment, val properties: Map<String, TomlValue>)
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:60]: > Potential memory leak: don't hold PsiElement, use SmartPsiElementPointer instead 

Let's ultimately get rid of this class and pass its values separately. Also, `properties` seems to be unused

**Review by ortem — CHANGES_REQUESTED** [review]
Currently, there are lots of fake git changes, so it's difficult to look into the real changes. To avoid these fake changes, let's separate `CratesLocalIndexService` refactoring into two commits:
* Extract `CratesLocalIndexServiceImpl` (`CratesLocalIndexService` class renaming)
* Add CratesLocalIndexService interface (new `CratesLocalIndexService` interface introducing)

File: toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt (hunk 9 above)
Comment (Kobzol) [reply toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:60]: I was already preparing a bit for other inspections, the Dependency code from this file will be probably refactored into a separate visitor to be used by multiple other inspections that might require the properties etc. But you are right that currently it's used only for this one inspection.

File: toml/src/main/resources/META-INF/toml-only.xml [hunk 10]
```diff
@@ -27,8 +27,15 @@
                          displayName="Cyclic feature dependency"
                          enabledByDefault="true" level="ERROR"
                          implementationClass="org.rust.toml.inspections.CargoTomlCyclicFeatureInspection"/>
+        <localInspection groupName="Rust"
+                         displayName="Crate not found"
+                         enabledByDefault="true" level="WARNING"
+                         implementationClass="org.rust.toml.inspections.CrateNotFoundInspection"/>
 
-        <applicationService serviceImplementation="org.rust.toml.crates.local.CratesLocalIndexService"/>
+        <applicationService
```
Comment (ortem) [inline toml/src/main/resources/META-INF/toml-only.xml:35]: Let's use the same formatting as for `CargoProjectsService`:
```
<applicationService serviceInterface="org.rust.toml.crates.local.CratesLocalIndexService"
                    serviceImplementation="org.rust.toml.crates.local.CratesLocalIndexServiceImpl"
                    testServiceImplementation="org.rust.toml.crates.local.TestCratesLocalIndexServiceImpl"/>
```

File: toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+            }
+        }
+        else -> null
+    }
+}
+
+private sealed class DependencyTableBase {
+    object DependencyTable : DependencyTableBase()
+    data class SpecificDependencyTable(val crate: TomlKeySegment, val crateName: String) : DependencyTableBase()
+}
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:84]: What do you think about this naming and fields order?
```suggestion
private sealed class DependencyTable {
    object General : DependencyTable()
    data class Specific(val crateName: String, val crateElement: TomlKeySegment) : DependencyTable()
}
```

File: toml/src/main/resources/META-INF/toml-only.xml [hunk 12]
```diff
@@ -27,8 +27,15 @@
                          displayName="Cyclic feature dependency"
                          enabledByDefault="true" level="ERROR"
                          implementationClass="org.rust.toml.inspections.CargoTomlCyclicFeatureInspection"/>
+        <localInspection groupName="Rust"
```
Comment (ortem) [inline toml/src/main/resources/META-INF/toml-only.xml:30]:
```suggestion
        <localInspection language="TOML" groupName="Rust"
```
it's just missed, I suppose? See `CargoTomlCyclicFeatureInspection` declaration above

File: toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+import org.rust.cargo.CargoConstants
+import org.rust.ide.experiments.RsExperiments
+import org.rust.openapiext.isFeatureEnabled
+import org.rust.toml.crates.local.CratesLocalIndexService
+import org.rust.toml.isDependencyKey
+import org.toml.lang.psi.*
+import org.toml.lang.psi.ext.TomlLiteralKind
+import org.toml.lang.psi.ext.kind
+
+class CrateNotFoundInspection : LocalInspectionTool() {
```
Comment (ortem) [inline toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:19]: I think this inspection should ignore dependencies with `path` or `git` specified? Seems like currently false-positive warning is shown instead. I mean something like `my_local_crate = { version = "1.0", path = "../my_local_crate" }` ([docs](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#multiple-locations))

**Review by ortem — CHANGES_REQUESTED** [review]

File: toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt (hunk 13 above)
Comment (ortem) [reply toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:19]: Also, looks like `TomlInlineTable` case is not covered by tests

Comment (Kobzol) [reply toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:19]: Well, I guess then it's back to having the `Dependency` class and parsing its properties? :sweat_smile: Otherwise I won't be able to check if it's a local/git dependency.

Comment (ortem) [reply toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:19]: Hm, probably 😄 
However please try to avoid holding `PsiElement`s. If it is necessary, `org.rust.openapiext.UtilsKt#createSmartPointer` might help to store a smart pointer instead

Comment (Kobzol) [reply toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:19]: Is it a problem in this case? I'm basically implementing a visitor which calls `visitDependency(object containing several PSI elements)`. It seems to me that it is the same situation as in normal PSI visitors.

Comment (ortem) [reply toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:20]: Oh, sorry, I looked at the code very briefly and didn't figure the problem out!
Looks like this warning is false-positive. I'll report this case to IntelliJ team a bit later. For now, you can either add `@Suppress("StatefulEp")` for `Dependency` class or rename it to something else (seems like it solves the warning), e.g. `DependencyCrate`

Comment (ortem) [reply toml/src/main/kotlin/org/rust/toml/inspections/CrateNotFoundInspection.kt:20]: BTW `DependencyCrate` and `DependencyTable` looks similar. Let's put them nearby and use the same naming for them, e.g. `val crateName: String, val crateElement: TomlKeySegment`.

`getDependencyTable` function could also be a factory method of `DependencyTable` named `from` or something like that, as far a I see.

**Review by ortem — APPROVED** [review]
Thank you!

bors r+

**Comment by Kobzol** [comment]
bors retry

**Comment by Kobzol** [comment]
bors retry

**Comment by Kobzol** [comment]
Thanks to @mibac138 for the groundwork!

**Comment by Kobzol** [comment]
@dima74 Maybe this shouldn't be put into v141 until https://github.com/intellij-rust/intellij-rust/issues/6812 is resolved.

**Comment by avrong** [comment]
@Kobzol discussed in private with @ortem. Decided that it's ok to have it now in the release as it under the experimental feature flag anyway. We will look into #6812 in details soon

**Comment by Kobzol** [comment]
Oh, I forgot that it's experimental. Then it's ok :)
